src/parser/SparqlParserHelpers.cpp

```diff
@@ -30,4 +36,73 @@ ParserAndVisitor::ParserAndVisitor(
     : ParserAndVisitor{std::move(input), disableSomeChecks} {
   visitor_.setPrefixMapManually(std::move(prefixes));
 }
+
+// _____________________________________________________________________________
+std::string ParserAndVisitor::unescapeUnicodeSequences(std::string input) {
+  std::u8string_view utf8View{std::bit_cast<char8_t*>(input.data()),
+                              input.size()};
+  std::string output;
+  size_t lastPos = 0;
+  UChar32 highSurrogate = 0;
+
+  auto throwError = [](bool condition, const std::string& message) {
+    if (!condition) {
+      throw InvalidSparqlQueryException{
+          absl::StrCat("Error in unicode escape sequence. ", message)};
+    }
+  };
+
+  constexpr size_t prefixLength = std::string_view{"\\U"}.size();
+
+  for (auto match :
+       ctre::search_all<R"(\\U[0-9A-Fa-f]{8}|\\u[0-9A-Fa-f]{4})">(utf8View)) {
+    output += input.substr(lastPos, match.data() - (utf8View.data() + lastPos));
+    lastPos = match.data() + match.size() - utf8View.data();
+
+    auto hexValue = match.to_view();
+
+    UChar32 codePoint;
+    auto* startPointer = std::bit_cast<const char*>(hexValue.data());
+    auto result =
+        std::from_chars(startPointer + prefixLength,
+                        startPointer + hexValue.size(), codePoint, 16);
+    AD_CORRECTNESS_CHECK(result.ec == std::errc{});
+
+    bool isFullCodePoint = hexValue.size() == 10;
+    AD_CORRECTNESS_CHECK(
```

Comment:
Add a short reference to some resource that explains the surrogate business in case somebody every wonders, and also state that this is necessary because htere is a compliance check.